drivers/lora/shell.c

```diff
@@ -0,0 +1,247 @@
+		shell_print(shell, DEFAULT_RADIO ":");
+		shell_print(shell, "  Frequency: %" PRIu32 " Hz",
+			    modem_config.frequency);
+		shell_print(shell, "  TX power: %" PRIi8 " dBm",
+			    modem_config.tx_power);
+		shell_print(shell, "  Bandwidth: %i kHz",
+			    bw_table[modem_config.bandwidth]);
+		shell_print(shell, "  Spreading factor: SF%i",
+			    (int)modem_config.datarate);
+		shell_print(shell, "  Coding rate: 4/%i",
+			    (int)modem_config.coding_rate + 4);
+		shell_print(shell, "  Preamble length: %" PRIu16,
+			    modem_config.preamble_len);
+
+		return 0;
+	}
+
+	for (int i = 1; i < argc; i += 2) {
+		char *param = argv[i];
+		char *eptr;
+		int value;
+
+		if (i + 1 >= argc) {
+			shell_error(shell, "'%s' expects an argument",
+				    argv[i]);
+			return -EINVAL;
+		}
+
+		value = strtol(argv[i + 1], &eptr, 0);
+		if (*eptr != '\0') {
+			shell_error(shell, "'%s' is not an integer",
+				    argv[i + 1]);
+			return -EINVAL;
+		}
+
+		if (!strcmp("freq", param)) {
+			modem_config.frequency = value;
+		} else if (!strcmp("tx-power", param)) {
+			modem_config.tx_power = value;
+		} else if (!strcmp("bw", param)) {
+			switch (value) {
+			case 125:
+				modem_config.bandwidth = BW_125_KHZ;
+				break;
+			case 250:
+				modem_config.bandwidth = BW_250_KHZ;
+				break;
+			case 500:
+				modem_config.bandwidth = BW_500_KHZ;
+				break;
+			default:
+				shell_error(shell, "Invalid bandwidth: %s",
+					    value);
+				return -EINVAL;
+			}
+		} else if (!strcmp("sf", param)) {
+			if (value < SF_6 || value > SF_12) {
+				shell_error(shell,
+					    "Invalid spreading factor: SF%i",
+					    value);
+				return -EINVAL;
+			}
+
+			modem_config.datarate = value;
+		} else if (!strcmp("cr", param)) {
+			if (value < 5 || value > 8) {
+				shell_error(shell, "Coding rate: 4/%i",
+					    value);
+				return -EINVAL;
+			}
+
+			modem_config.coding_rate = CR_4_5 + value - 5;
+		} else if (!strcmp("pre-len", param)) {
+			modem_config.preamble_len = value;
+		} else {
+			shell_error(shell, "Unknown parameter '%s'", param);
+			return -EINVAL;
+		}
+	}
+
+	return 0;
+}
+
+static int cmd_lora_send(const struct shell *shell,
+			size_t argc, char **argv)
+{
+	int ret;
+	struct device *dev;
+
+	modem_config.tx = true;
+	dev = get_modem(shell, &modem_config);
+	if (!dev) {
+		return -ENODEV;
+	}
+
+	ret = lora_send(dev, argv[1], strlen(argv[1]));
+	if (ret < 0) {
+		shell_error(shell, "LoRa send failed: %i", ret);
+		return ret;
+	}
+
+	return 0;
+}
+
+static int cmd_lora_recv(const struct shell *shell,
+			size_t argc, char **argv)
+{
+	static char buf[0xff];
+	struct device *dev;
+	int ret;
+	s16_t rssi;
+	s8_t snr;
+
+	modem_config.tx = false;
+	dev = get_modem(shell, &modem_config);
+	if (!dev) {
+		return -ENODEV;
+	}
+
+	ret = lora_recv(dev, buf, sizeof(buf), K_FOREVER, &rssi, &snr);
+	if (ret < 0) {
+		shell_error(shell, "LoRa recv failed: %i", ret);
+		return ret;
+	}
+
+	shell_print(shell, "Data: %s", buf);
```

Comment:
We should use `shell_hexdump()` here for data printing. One thing that I noticed is that `shell_hexdump()` misses printing bytes as characters (in case buffer contains valid string), which is nicely done in `LOG_HEXDUMP_*()`.

Reply:
Hmm, this is a good point. It's unfortunately a choice between three bad options, shell_hexdump that doesn't play nicely with send (that sends a string), LOG_HEXDUMP that doesn't play nicely with the shell, and shell_hexdump that doesn't print binary. I think I prefer the current option for now, but it should be changed to shell_hexdump when/if it prints printable characters.